docs/observability.md

```diff
@@ -0,0 +1,69 @@
+- Receiving data rate, broken down by receivers and by data type (traces/metrics).
+- Exporting data rate, broken down by exporters and by data type (traces/metrics).
+- Data drop rate due to throttling, broken down by data type.
+- Data drop rate due to invalid data received, broken down by data type.
+- Current throttling state: Not Throttled/Throttled by Downstream/Internally Saturated.
+- Incoming connection count, broken down by receiver.
+- Incoming connection rate (new connections per second), broken down by receiver.
+- In-memory queue size (in bytes and in units).
+- Persistent queue size (when supported).
+- End-to-end latency (from receiver input to exporter output). Note that with multiple receivers/exporters we potentially have NxM data paths, each with different latency (plus different pipelines in the future), so realistically we should likely expose the average of all data paths (perhaps broken down by pipeline).
+- Latency broken down by pipeline elements (including exporter network roundtrip latency for request/response protocols).
+
+“Rate” values must reflect the average rate of the last 10 seconds. Rates must exposed in bytes/sec and units/sec (e.g. spans/sec).
+
+Note: some of the current values may be calculated as derivatives of cumulative values in the backend, so it is an open question if we want to expose them separately or no.
+
+### Cumulative Values
+
+- Total received data, broken down by receivers and by data type (traces/metrics)
+- Total exported data, broken down by exporters and by data type (traces/metrics)
+- Total dropped data due to throttling, broken down by data type.
+- Total dropped data due to invalid data received, broken down by data type.
+- Total incoming connection count, broken down by receiver.
+- Uptime since start.
+
+### Aggregate Values
+
+For the following metrics we want to track min, average and max since start: TBD.
+
+### Trace or Log on Events
+
+We want to generate the following events (log and/or send as a trace with additional data):
+
+- Service started/stopped.
+- Service reconfigured (if we support on-the-fly reconfiguration).
+- Begin dropping due to throttling (include throttling reason, e.g. local saturation, downstream saturation, downstream unavailable, etc).
+- Stop dropping due to throttling.
+- Begin dropping due to invalid data (include sample/first invalid data).
+- Stop dropping due to invalid data.
+- Crash detected (differentiate clean stopping and crash, possibly include crash data if available).
+
+For begin/stop events we need to define an appropriate hysteresis to avoid generating too many events. Note that begin/stop events cannot be detected in the backend simply as derivatives of current rates, the events include additional data that is not present in the current value.
+
+## How We Expose Metrics/Traces
```

Comment:
Consider adding a section on how to enable/disable self-monitored metrics/traces.

Reply:
This is a vision document, that only describes goals. The exact implementation will need to deal with the question "how". I suggest that it is done later when we actually implement this vision or parts of it.